tests/80torture/20json.pl

```diff
@@ -27,3 +27,51 @@
          )->followed_by( \&main::expect_http_400 ),
       );
    };
+
+# Floats (including NaN, Infinity, and -Infinity) should be rejected.
+test "Invalid JSON floats",
+   requires => [ local_user_and_room_fixtures(
+      room_opts => { room_version => "org.matrix.strict_canonicaljson" }
+   ), ],
+
+   do => sub {
+      my ( $user, $room_id ) = @_;
+
+      Future->needs_all(
+         do_request_json_for( $user,
+            method  => "POST",
+            uri     => "/r0/rooms/$room_id/send/sytest.dummy",
+            content => {
+               msgtype => "sytest.dummy",
+               body    => 1.1,
+            },
+         )->followed_by( \&main::expect_http_400 ),
+
+         do_request_json_for( $user,
+            method  => "PUT",
+            uri     => "/r0/rooms/$room_id/send/sytest.dummy",
+            content => {
+               msgtype => "sytest.dummy",
+               body    => "NaN",
+            },
+         )->followed_by( \&main::expect_http_400 ),
+
+         do_request_json_for( $user,
+            method  => "PUT",
+            uri     => "/r0/rooms/$room_id/send/sytest.dummy",
+            content => {
+               msgtype => "sytest.dummy",
+               body    => "inf",
```

Comment:
This produces a float instead of a string?

Reply:
I'm fairly certain this boils down to a "perl is weird", but I'll double check that this is doing what I expect.

Reply:
OK, so the tests were broken and not testing what I thought they were... so yeah this is wrong. 😢

Reply:
I fixed this in 972e60bac09c1fc4088d4c14d778f1e987aceaa7, but it sends the following JSON now:

```json
{"body":-nan,"msgtype":"sytest.dummy"}
```

Which is invalid, but in a slightly different way than I expected. I think it is OK to leave these tests, but I'm not sure if they really add value beyond the standard float one. Any thoughts?

Reply:
Probably nice to still keep, but I would leave a comment explaining that it doesn't produce valid JSON but still serves to test validity code?

Reply:
I decided to rewrite these a bit and add some of the Python values which could cause issues. Going to ask you to look at this again, sorry!